nav2_system_tests/src/behaviors/assisted_teleop/assisted_teleop_behavior_tester.cpp

```diff
@@ -0,0 +1,280 @@
+// limitations under the License. Reserved.
+
+#include <string>
+#include <random>
+#include <tuple>
+#include <memory>
+#include <iostream>
+#include <chrono>
+#include <sstream>
+#include <iomanip>
+
+#include "assisted_teleop_behavior_tester.hpp"
+#include "nav2_util/geometry_utils.hpp"
+
+using namespace std::chrono_literals;
+using namespace std::chrono;  // NOLINT
+
+namespace nav2_system_tests
+{
+
+AssistedTeleopBehaviorTester::AssistedTeleopBehaviorTester()
+: is_active_(false),
+  initial_pose_received_(false)
+{
+  node_ = rclcpp::Node::make_shared("assisted_teleop_behavior_test");
+
+  tf_buffer_ = std::make_shared<tf2_ros::Buffer>(node_->get_clock());
+  tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);
+
+  client_ptr_ = rclcpp_action::create_client<AssistedTeleop>(
+    node_->get_node_base_interface(),
+    node_->get_node_graph_interface(),
+    node_->get_node_logging_interface(),
+    node_->get_node_waitables_interface(),
+    "assisted_teleop");
+
+  initial_pose_pub_ =
+    node_->create_publisher<geometry_msgs::msg::PoseWithCovarianceStamped>("initialpose", 10);
+
+  preempt_pub_ =
+    node_->create_publisher<std_msgs::msg::Empty>("preempt_teleop", 10);
+
+  cmd_vel_pub_ =
+    node_->create_publisher<geometry_msgs::msg::Twist>("teleop_cmd_vel", 10);
+
+  subscription_ = node_->create_subscription<geometry_msgs::msg::PoseWithCovarianceStamped>(
+    "amcl_pose", rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
+    std::bind(&AssistedTeleopBehaviorTester::amclPoseCallback, this, std::placeholders::_1));
+
+  filtered_vel_sub_ = node_->create_subscription<geometry_msgs::msg::Twist>(
+    "cmd_vel",
+    rclcpp::SystemDefaultsQoS(),
+    std::bind(&AssistedTeleopBehaviorTester::filteredVelCallback, this, std::placeholders::_1));
+
+  std::string costmap_topic = "/local_costmap/costmap_raw";
+  std::string footprint_topic = "/local_costmap/published_footprint";
+
+  costmap_sub_ = std::make_shared<nav2_costmap_2d::CostmapSubscriber>(
+    node_,
+    costmap_topic);
+
+  footprint_sub_ = std::make_shared<nav2_costmap_2d::FootprintSubscriber>(
+    node_,
+    footprint_topic,
+    *tf_buffer_);
+
+  collision_checker_ = std::make_unique<nav2_costmap_2d::CostmapTopicCollisionChecker>(
+    *costmap_sub_,
+    *footprint_sub_
+  );
+
+  stamp_ = node_->now();
+}
+
+AssistedTeleopBehaviorTester::~AssistedTeleopBehaviorTester()
+{
+  if (is_active_) {
+    deactivate();
+  }
+}
+
+void AssistedTeleopBehaviorTester::activate()
+{
+  if (is_active_) {
+    throw std::runtime_error("Trying to activate while already active");
+    return;
+  }
+
+  while (!initial_pose_received_) {
+    RCLCPP_WARN(node_->get_logger(), "Initial pose not received");
+    sendInitialPose();
+    std::this_thread::sleep_for(100ms);
+    rclcpp::spin_some(node_);
+  }
+
+  // Wait for lifecycle_manager_navigation to activate behavior_server
+  std::this_thread::sleep_for(10s);
```

Comment:
You've got alot of sleeps in this test... is there a way to reduce these times or use other utilities to be event based (e.g. check the lifecycle status of it)? All of these waits _kill_ us in our CI times as they add up

Reply:
OK - well try to minimize these or at least time how long it should practically take and set the timeouts to if that some leniency. If we set these to really high values for something that is relatively quick, then that just adds to CI time